File: Salted Kalman Filter/Matlab/SKF.m
classdef SKF
    properties
        current_state        % Current state
        current_cov          % Current covariance matrix
        current_mode         % Current mode of the system
        dt                   % Time step
        noise_matrices_dict  % Noise matrices for each mode
        dynamics_dict        % Dynamics functions for each mode
        resets_dict          % Reset functions for each allowable transition
        guards_dict          % Guard functions for each allowable transition
        parameters           % System parameters
        n_states             % Number of states
    end
    
    methods
        function obj = SKF(init_state, init_mode, init_cov, dt, noise_matrices, dynamics, resets, guards, parameters)
            % Constructor
            obj.current_state = init_state;
            obj.current_cov = init_cov;
            obj.current_mode = init_mode;
            obj.dt = dt;
            obj.noise_matrices_dict = noise_matrices;
            obj.dynamics_dict = dynamics;
            obj.resets_dict = resets;
            obj.guards_dict = guards;
            obj.parameters = parameters;
            obj.n_states = length(init_state);
        end
        
        function [obj, current_state, current_cov] = predict(obj, current_time, inputs)
            % Perform the prediction (prior update) step
            end_time = current_time + obj.dt;
            
            % Set up dynamics function
            current_dynamics = solve_ivp_dynamics_func( ...
                obj.dynamics_dict, obj.current_mode, inputs, obj.dt, obj.parameters);
            % [current_guards, possible_modes] = solve_ivp_guard_funcs( ...
                % obj.guards_dict, obj.current_mode, inputs, obj.dt, obj.parameters);
            [current_guards, possible_modes] = obj.evaluateGuards(inputs);

            % Perform integration
            sol = ode45(current_dynamics, [current_time, end_time], obj.current_state, odeset('Events', current_guards));
            current_state = sol.y(:, end);
            
            % Check for hybrid events and apply reset if guard was hit
            [hybrid_event_state, hybrid_event_time, new_mode] = solve_ivp_extract_hybrid_events(sol, possible_modes);

            while ~isempty(new_mode)
                % Apply reset
                current_state = obj.resets_dict.(obj.current_mode).(new_mode).r( ...
                    hybrid_event_state, inputs, obj.dt, obj.parameters);
                
                % Apply covariance updates (dynamics and saltation matrix)
                dynamics_cov = obj.dynamics_dict.(obj.current_mode).A_disc( ...
                    obj.current_state, inputs, sol.x(end) - sol.x(1), obj.parameters);
                obj.current_cov = dynamics_cov * obj.current_cov * dynamics_cov' ...
                    + obj.noise_matrices_dict.(obj.current_mode).W;

                salt = compute_saltation_matrix( ...
                    hybrid_event_time, hybrid_event_state, inputs, obj.dt, obj.parameters, ...
                    obj.current_mode, new_mode, obj.dynamics_dict, obj.resets_dict, obj.guards_dict, current_state);
                obj.current_cov = salt * obj.current_cov * salt';
                
                % Update guard and re-simulate
                obj.current_mode = new_mode;
                current_dynamics = solve_ivp_dynamics_func( ...
                    obj.dynamics_dict, obj.current_mode, inputs, obj.dt, obj.parameters);
                % [current_guards, possible_modes] = solve_ivp_guard_funcs( ...
                    % obj.guards_dict, obj.current_mode, inputs, obj.dt, obj.parameters);
                [current_guards, possible_modes] = evaluateGuards(obj, inputs);
                sol = ode45(current_dynamics, [hybrid_event_time, end_time], current_state, odeset('Events', current_guards));
                [hybrid_event_state, hybrid_event_time, new_mode] = solve_ivp_extract_hybrid_events(sol, possible_modes);
            end

            % Apply final covariance update
            dynamics_cov = obj.dynamics_dict.(obj.current_mode).A_disc( ...
                obj.current_state, inputs, sol.x(end) - sol.x(1), obj.parameters);
            obj.current_cov = dynamics_cov * obj.current_cov * dynamics_cov' + obj.noise_matrices_dict.(obj.current_mode).W;
            
            % Update state and return
            obj.current_state = current_state;
            current_state = obj.current_state;
            current_cov = obj.current_cov;
        end
        
        function [obj, current_state, current_cov] = update(obj, current_time, current_input, measurement)
            % Perform the update (posterior update) step
            C = obj.dynamics_dict.(obj.current_mode).C(obj.current_state, obj.parameters);
            V = obj.noise_matrices_dict.(obj.current_mode).V;
            K = obj.current_cov * C' / (C * obj.current_cov * C' + V);

            % Measurement update
            measurement_est = obj.dynamics_dict.(obj.current_mode).y(obj.current_state, obj.parameters);
            residual = measurement - measurement_est;
            obj.current_state = obj.current_state + K * residual;
            obj.current_cov = (eye(obj.n_states) - K * C) * obj.current_cov;

            % Check guard conditions and apply reset if needed
            % [current_guards, possible_modes] = solve_ivp_guard_funcs(obj.guards_dict, obj.current_mode, current_input, obj.dt, obj.parameters);
            [current_guards, possible_modes] = evaluateGuards(obj, current_input);
            guard_names = fieldnames(obj.guards_dict.(obj.current_mode));
            for guard_idx = 1:length(current_guards)
                guard_name = guard_names{guard_idx};
                if current_guards(current_time, obj.current_state) < 0
                    % Apply reset due to guard condition
                    new_mode = possible_modes(guard_idx);
                    new_state = obj.resets_dict.(obj.current_mode).(new_mode).r(obj.current_state, current_input, obj.dt, obj.parameters);
                    
                    % Update covariance with saltation matrix
                    salt = compute_saltation_matrix( ...
                        current_time, obj.current_state, current_input, obj.dt, obj.parameters, ...
                        obj.current_mode, new_mode, obj.dynamics_dict, obj.resets_dict, obj.guards_dict, new_state);
                    obj.current_state = new_state;
                    obj.current_cov = salt * obj.current_cov * salt';
                    break;
                end
            end

            % Return updated state and covariance
            current_state = obj.current_state;
            current_cov = obj.current_cov;
        end
    end

    methods (Access = private)
    function [guards, possible_modes] = evaluateGuards(obj, inputs)
        % Generate guard functions and possible modes for the current mode
        [guard_funcs, possible_modes] = solve_ivp_guard_funcs(obj.guards_dict, obj.current_mode, inputs, obj.dt, obj.parameters);
        
        % Combine individual guard functions into a single function handle
        % compatible with ode45's event detection requirements
        guards = @(t, states) obj.evaluateGuardsForODE(guard_funcs, t, states);
    end
    
    function [value, isterminal, direction] = evaluateGuardsForODE(obj, guard_funcs, t, states)
        % Evaluate each guard function and return in a format compatible with ode45
        value = cellfun(@(g) g(t, states), guard_funcs); % Evaluate all guard functions
        isterminal = ones(size(value));    % Stop integration if any guard is triggered
        direction = -ones(size(value));    % Detect zero crossings in the decreasing direction
    end

    function [new_state, new_mode] = applyReset(obj, impact_state, inputs, new_mode)
        % Apply the reset function and return new state and mode
        reset_func = obj.resets_dict.(obj.current_mode).(new_mode).r;
        new_state = reset_func(impact_state, inputs, obj.dt, obj.parameters);
    end
    end
end

File: Salted Kalman Filter/Python/src/skf.py
"""
skf.py

This module implements the Salted Kalman Filter (SKF) for hybrid (mode-switching) dynamical systems.
The SKF extends standard Kalman filtering to handle hybrid transitions by incorporating saltation matrices 
to correctly propagate covariance across discrete resets.

Key Features:
- Continuous-time state prediction via ODE integration with discrete event detection.
- State and covariance propagation across hybrid transitions using saltation matrices.
- Measurement update step with standard Kalman filter correction.
- Hybrid posterior updates when measurements indicate mode transitions.

Main Class:
- SKF:
    - predict: Performs a prior update (state and covariance prediction) over one timestep, handling hybrid transitions.
    - update: Performs a posterior update using a new noisy measurement and adjusts state/covariance if mode transitions occur.
"""

import sys
import pathlib
import numpy as np
from scipy.integrate import solve_ivp

sys.path.append(str(pathlib.Path(__file__).parent.parent))
from src.hybrid_helper_functions import (
    solve_ivp_dynamics_func,
    solve_ivp_guard_funcs,
    solve_ivp_extract_hybrid_events,
    compute_saltation_matrix,
)

class SKF:
    def __init__(
        self,
        init_state,
        init_mode,
        init_cov,
        dt,
        noise_matrices,
        dynamics,
        resets,
        guards,
        parameters,
    ):
        """
        init_state (np.array): Initial state.
        init_cov (np.array): Initial covariance.
        noise_matrices (np.array): Noise matrices for each mode.
        dynamics (dict): Dynamics for each mode.
        resets (dict): Resets for each allowable transition.
        guards (dict): Guards for each allowable transition.
        parameters (np.array): Extra parameters of the system.
        """
        self._current_state = init_state
        self._current_cov = init_cov
        self._current_mode = init_mode
        self._dt = dt
        self._noise_matrices_dict = noise_matrices
        self._dynamics_dict = dynamics
        self._resets_dict = resets
        self._guards_dict = guards
        self._parameters = parameters

        self._n_states = np.shape(self._current_state)[0]

    def predict(self, current_time, inputs):
        """
        Prior update.
        """
        end_time = current_time + self._dt

        """ Integrate for dt. """
        current_dynamics = solve_ivp_dynamics_func(
            self._dynamics_dict, self._current_mode, inputs, self._dt, self._parameters
        )
        current_guards, possible_modes = solve_ivp_guard_funcs(
            self._guards_dict, self._current_mode, inputs, self._dt, self._parameters
        )

        current_start_state = self._current_state.copy()
        sol = solve_ivp(
            current_dynamics,
            [current_time, end_time],
            self._current_state,
            events=current_guards,
        )
        current_state = np.zeros(self._n_states)

        """ If we hit guard, apply reset. """
        (
            hybrid_event_state,
            hybrid_event_time,
            new_mode,
        ) = solve_ivp_extract_hybrid_events(sol, possible_modes)

        while new_mode is not None:
            """Apply reset."""
            current_state = self._resets_dict[self._current_mode][new_mode]['r'](
                hybrid_event_state, inputs, self._dt, self._parameters
            ).reshape(np.shape(hybrid_event_state))

            """ Apply covariance updates: dynamics and saltation matrix."""
            dynamics_cov = self._dynamics_dict[self._current_mode]["A_disc"](
                current_start_state, inputs, sol.t[-1] - sol.t[0], self._parameters
            )
            self._current_cov = (
                dynamics_cov @ self._current_cov @ dynamics_cov.T
                + self._noise_matrices_dict[self._current_mode]["W"]
            )
            salt = compute_saltation_matrix(
                t=hybrid_event_time,
                pre_event_state=hybrid_event_state,
                inputs=inputs,
                dt=self._dt,
                parameters=self._parameters,
                pre_mode=self._current_mode,
                post_mode=new_mode,
                dynamics_dict=self._dynamics_dict,
                resets_dict=self._resets_dict,
                guards_dict=self._guards_dict,
                post_event_state=current_state,
            )
            self._current_cov = salt @ self._current_cov @ salt.T

            """ Update guard and simulate. """
            self._current_mode = new_mode
            current_dynamics = solve_ivp_dynamics_func(
                self._dynamics_dict,
                self._current_mode,
                inputs,
                self._dt,
                self._parameters,
            )
            current_guards, possible_modes = solve_ivp_guard_funcs(
                self._guards_dict,
                self._current_mode,
                inputs,
                self._dt,
                self._parameters,
            )
            current_start_state = current_state.copy()
            sol = solve_ivp(
                current_dynamics,
                [hybrid_event_time, end_time],
                current_state,
                events=current_guards,
            )
            (
                hybrid_event_state,
                hybrid_event_time,
                new_mode,
            ) = solve_ivp_extract_hybrid_events(sol, possible_modes)

        """ Once no more hybrid events, grab the terminal states. """
        for idx in range(len(sol.y)):
            """Grab the state at the last timestep."""
            current_state[idx] = sol.y[idx][-1]

        """ Propagate the rest of the covariance. """
        dynamics_cov = self._dynamics_dict[self._current_mode]["A_disc"](
            current_start_state, inputs, sol.t[-1] - sol.t[0], self._parameters
        )
        
        self._current_cov = (
            dynamics_cov @ self._current_cov @ dynamics_cov.T
            + self._noise_matrices_dict[self._current_mode]["W"]
        )

        self._current_state = current_state
        return self._current_state, self._current_cov

    def update(self, current_time, current_input, measurement):
        """
        Posterior update.
        When a new measurement comes in, update the covariance.
        If updated state is pulled into new mode, then apply saltation matrix and reset.
        """
        C = self._dynamics_dict[self._current_mode]['C'](
                self._current_state,
                self._parameters,
            )
        V = self._noise_matrices_dict[self._current_mode]['V']
        K = self._current_cov@C.T@np.linalg.inv(C@self._current_cov@C.T + V)

        """ Measurement update. """
        measurement_est = self._dynamics_dict[self._current_mode]['y'](
                self._current_state,
                self._parameters,
            ).flatten()
        C = self._dynamics_dict[self._current_mode]['C'](
                self._current_state,
                self._parameters,
            )
        residual = measurement - measurement_est
        self._current_state = self._current_state + K@residual
        self._current_cov = self._current_cov - K@C@self._current_cov

        """ Check guard conditions. If any guard has been reached, then apply hybrid posterior update. """
        current_guards, possible_modes = solve_ivp_guard_funcs(
            self._guards_dict, self._current_mode, current_input, self._dt, self._parameters
        )
        for guard_idx in range(len(current_guards)):
            if current_guards[guard_idx](current_time, self._current_state) < 0:
                new_mode = possible_modes[guard_idx]
                """Apply reset."""
                new_state = self._resets_dict[self._current_mode][new_mode]['r'](
                    self._current_state, current_input, self._dt, self._parameters
                ).reshape(np.shape(self._current_state))

                """ Apply covairance updates: dynamics and saltation matrix."""
                salt = compute_saltation_matrix(
                    t=current_time,
                    pre_event_state=self._current_state,
                    inputs=current_input,
                    dt=self._dt,
                    parameters=self._parameters,
                    pre_mode=self._current_mode,
                    post_mode=new_mode,
                    dynamics_dict=self._dynamics_dict,
                    resets_dict=self._resets_dict,
                    guards_dict=self._guards_dict,
                    post_event_state=new_state,
                )
                self._current_state = new_state
                self._current_cov = salt @ self._current_cov @ salt.T
                break

        return self._current_state, self._current_cov
